Implement Meson (take 2)

Kangie opened the pull request:
This PR picks up after https://github.com/c3lang/c3c/pull/1097 (thanks @Opaatia) and rebases it on current HEAD.

I was asked to look into completing the port, so I'm not super familiar with c3c - apologies for any obvious oversights / questions!

While some existing CMake logic remains to be ported, there's an opportunity to potentially simplify it. Any guidance on the preferred approach (simplifying vs. following existing logic) would be greatly appreciated.

### Progress and remaining tasks:

Done so far:

- [x] remove tilde-backend sources completely including `.gitmodules`
- [x] generate `git_hash.h` and conifgure `git_hash.h.in` (new file)
- [x] set default project arguments `c{,pp}_std`, `warning_level` `buildtype`
- [x] add summary output
- [x] llvm rtti detection
- [x] Wired up CI for Linux (ubuntu-latest), MacOS, msys2, Windows.
- [x] Docs (patches didn't apply cleanly; just focus on meson for now)
- [x] LLVM is again optional
- [x] MacOS support (pending CI testing)
- [x] Windows/msvc support (pending CI testing)  

Not implemented:

- Fetching LLVM on Windows
- Mimalloc (would need a submodule, or ideally a [Meson WrapDB](https://github.com/mesonbuild/wrapdb)

Please let me know if there are any changes that I can make while I'm here.

lerno commented:
I think LLVM optional was for people working on additional backends. Normally, LLVM should be required.

lerno commented:
Let me know what else I can help out with.

Kangie commented:
Excellent, thank you. I can get that logic in re: LLVM.

Re `tilde-backend` - I note that a lot of the sources have in-repo been removed, but there's still a `.gitmodules` referencing the repo that doesn't properly init:

https://github.com/c3lang/c3c/blob/70da1f748a3d3d7195830c6e18967d875bff6c2b/.gitmodules#L1-L3

Should I fix the submodule or purge the `.gitmodules`?

lerno commented:
You can drop .gitmodules completely.

Kangie commented:
Added some CI that will let me get started on 'Windows' and 'MacOS'. Hopefully it's all straightforward!

Rebased on current `dev`. Test suite mostly running via `meson` on Linux, and what isn't running I'll look into - I may have fat-fingered some unit tests while removing boilerplate.

```
Ok:                 388 
Expected Fail:      0   
Fail:               13  
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   
```

and 

```
================================== 203/401 ===================================
test:         Test Suite - errors - simple_static_failable
start time:   13:56:29
duration:     1.00s
result:       exit status 0
command:      ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 MALLOC_PERTURB_=70 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 /usr/bin/python3 ../test/src/check.py ../test/src/tester.py ./c3c ../test/test_suite/errors/simple_static_failable.c3t
==============================================================================

================================== 204/401 ===================================
test:         Test Suite - expressions - rt_float
start time:   13:56:30
duration:     0.18s
result:       exit status 1
command:      MALLOC_PERTURB_=200 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 /usr/bin/python3 ../test/src/check.py ../test/src/tester.py ./c3c ../test/test_suite/expressions/rt_float.c3t
----------------------------------- stdout -----------------------------------
Error running tester.py: Command '['python3', '../test/src/tester.py', './c3c', '../test/test_suite/expressions/rt_float.c3t']' returned non-zero exit status 255.
==============================================================================

================================== 205/401 ===================================
test:         Test Suite - errors - try_with_chained_unwrap
start time:   13:56:29
duration:     0.96s
result:       exit status 0
command:      MALLOC_PERTURB_=209 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 /usr/bin/python3 ../test/src/check.py ../test/src/tester.py ./c3c ../test/test_suite/errors/try_with_chained_unwrap.c3t
==============================================================================
```

Kangie commented:
I've squashed down the previous PR into a single commit, and my changes so far down as far as I can.

May need to script up something to generate the list of tests to be 100% sure I got 'em all but we're looking pretty good.

If someone is able to unblock the CI workflow I'll see if I can get Windows and MacOS builds going - Pretty much all of the logic is ported and we're just on to working out what I broke in CI and any edge cases or things I overlooked :)

lerno commented:
There are two sets of tests: one is the unit test variant, the other is what runs by way of a python script. 

The python script accepts two types of files (from the ancient times when the compiler was first written): The regular .c3 files which are just compiled normally, doing type checking, and the .c3t files which actually compile things and check the LLVM IR output. They have slightly different directives (in comments) available to them, and because originally things like `/*` without terminator `*/` wasn't available, it needed to be in a slightly different format. But actually this is not strictly needed anymore, so when the c3-based external tester is done, these could also be `.c3` files. (Although it's been kind of useful to see which ones are supposed to error and which ones are supposed to compile).

lerno commented:
Still some errors I'm afraid.

Kangie commented:
> Still some errors I'm afraid.

Yes, I'm steadily working my way through them - I might work off a different branch so that I stop force pushing and spamming here :)

I have gotten to the point where compilation succeeds for MacOS, Linux, and Windows.

I'll get the CI stuff ready for review and ping you then - I expect that I'll need to make some changes once it's working!

Current notes / questions:

- I note that ubuntu-latest was changed to an older ubuntu, I assume because of the same issues I ran into with clang 17,
- I've swtched to shell to `pwsh` from `cmd` for windows to take advantage of modern features and syntax.)
- How far back do we want Meson CI running? I haven't targeted 'old ubuntu' (20.04) but there's no reason that it won't work
- I never asked (just focused on completing the port) but is the intent to run this alongside CMake or replace CMake?

lerno commented:
My understanding is that with Meson it's possible to work with distributions that require something like Meson to build.

ThomasAdam commented:
@lerno,

Yes -- and I would suggest that in light of the work being done on meson here, that this subsumes CMake directly, otherwise you're maintaining two buildsystems needlessly, IMO.

Additionally, for those people who might not have python3 available, etc., or have a severe allergic reaction to it, `muon` exists, written in C.

Kangie commented:
Still working on this, I think it accidentally got closed when `dev` was deleted as I rebasing off that branch!

I'll continue working on the CI side in my fork until I'm satisfied with it (we're pretty close, may have to spin up a windows VM to get across the finer points of meson llvm detection on that platform). In terms of general usability it builds on Windows, Linux, and MacOS with msvc and clang respectively (maybe GCC for Linux too, can't remember if I tested that).
